feat: add C ndarray interface and refactor implementation for `stats/base/dmskrange`

aayush0325 opened the pull request:
Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

-   adds C ndarray interface for `stats/base/dmskrange`
-   refactors JavaScript implementation
-   updates tests, docs, examples and benchmarks

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aayush0325 committed:
feat: add C ndarray interface and refactor implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 commented:
same comment regarding line length exceeding in `lib/ndarray.native.js`, have added a TODO for that for the reviewers convenience

aayush0325 committed:
chore: include stdint.h in header file as uint8 is used

aayush0325 committed:
chore: fix copy paste error

aayush0325 committed:
chore: add missing const

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/lib/ndarray.native.js:
```diff
@@ -42,27 +40,16 @@ var addon = require( './dmskrange.native.js' );
-	viewM = new Uint8Array( mask.buffer, mask.byteOffset+(mask.BYTES_PER_ELEMENT*offsetMask), mask.length-offsetMask ); // eslint-disable-line max-len
-	return addon( N, viewX, strideX, viewM, strideMask );
+	// TODO - add eslint comment to escape max line length warning(if needed)
+	return addon.ndarray( N, x, strideX, offsetX, mask, strideMask, offsetMask );
```
@aayush0325 The necessary comment was present in the lines you removed. See above in the diff.

```suggestion
	return addon.ndarray( N, x, strideX, offsetX, mask, strideMask, offsetMask ); // eslint-disable-line max-len
```

aayush0325 replied:
ah, i'll just push this wherever it is needed. thanks for pointing it out

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
```diff
@@ -184,6 +176,123 @@ console.log( v );
+
+#### stdlib_strided_dmskrange( N, \*X, strideX, Mask, strideMask )
+
+Computes the [range][range] of a double-precision floating-point strided array `x` according to a `mask`.
```
```suggestion
Computes the [range][range] of a double-precision floating-point strided array according to a mask.
```

In this case, the text variables don't match the signature in the line above. We can either use `X` and `Mask` or just remove. I am fine removing here. The main thing is that we don't want to be inconsistent in the immediate vicinity of the signature (e.g., `X` vs `x`, etc).

aayush0325 replied:
noted, taking care of this everywhere else

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
```diff
@@ -184,6 +176,123 @@ console.log( v );
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **X**: `[in] double*` input array.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+-   **Mask**: `[in] uint_8*`. If a `mask` array element is `0`, the corresponding element in `x` is considered valid and included in computation. If a `mask` array element is `1`, the corresponding element in `x` is considered invalid/missing and excluded from computation.
```
```suggestion
-   **Mask**: `[in] uint_8*` mask array. If a `Mask` array element is `0`, the corresponding element in `X` is considered valid and included in computation. If a `Mask` array element is `1`, the corresponding element in `X` is considered invalid/missing and excluded from computation.
```

kgryte replied:
Always include an initial lowercase description before the more extended description, as done in L61.

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
````diff
@@ -184,6 +176,123 @@ console.log( v );
+double stdlib_strided_dmskrange( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const uint8_t *Mask, const CBLAS_INT strideMask );
+```
+
+#### stdlib_strided_dmskrange_ndarray( N, \*X, strideX, offsetX )
````
Missing parameters.

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
```diff
@@ -184,6 +176,123 @@ console.log( v );
+
+#### stdlib_strided_dmskrange_ndarray( N, \*X, strideX, offsetX )
+
+Computes the [range][range] of a double-precision floating-point strided array according to a `mask` and using alternative indexing semantics.
```
```suggestion
Computes the [range][range] of a double-precision floating-point strided array according to a mask and using alternative indexing semantics.
```

Consistency.

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
```diff
@@ -184,6 +176,123 @@ console.log( v );
+#include <stdint.h>
+
+const double x[] = { 1.0, -2.0, 2.0 };
+const uint_8 mask = { 0, 1, 0 };
```
```suggestion
const uint_8 mask[] = { 0, 1, 0 };
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
```diff
@@ -184,6 +176,123 @@ console.log( v );
+#include <stdint.h>
+
+const double x[] = { 1.0, -2.0, 2.0 };
+const uint_8 mask = { 0, 1, 0 };
```
```suggestion
const uint_8 mask[] = { 0, 1, 0 };
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
```diff
@@ -184,6 +176,123 @@ console.log( v );
+-   **X**: `[in] double*` input array.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
+-   **Mask**: `[in] uint_8*`. If a `mask` array element is `0`, the corresponding element in `x` is considered valid and included in computation. If a `mask` array element is `1`, the corresponding element in `x` is considered invalid/missing and excluded from computation.
```
```suggestion
-   **Mask**: `[in] uint_8*` mask array. If a `Mask` array element is `0`, the corresponding element in `X` is considered valid and included in computation. If a `Mask` array element is `1`, the corresponding element in `X` is considered invalid/missing and excluded from computation.
```

We need to be consistent with capitalization. If the parameter is capitalized, that needs to be reflected in descriptions.

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
````diff
@@ -184,6 +176,123 @@ console.log( v );
+-   **offsetMask**: `[in] CBLAS_INT` starting index for `Mask`.
+
+```c
+double stdlib_strided_dmskrange_ndarray( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, const uint8_t *Mask, const CBLAS_INT strideMask , const CBLAS_INT offsetMask );
````
```suggestion
double stdlib_strided_dmskrange_ndarray( const CBLAS_INT N, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, const uint8_t *Mask, const CBLAS_INT strideMask, const CBLAS_INT offsetMask );
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
```diff
@@ -184,6 +176,123 @@ console.log( v );
+    const int strideX = 2;
+    const int strideMask = 2;
+
+    // Compute the minimum value:
```
```suggestion
    // Compute the range:
```

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
```diff
@@ -184,6 +176,123 @@ console.log( v );
+    double v = stdlib_strided_dmskrange( N, x, strideX, mask, strideMask );
+
+    // Print the result:
+    printf( "min: %lf\n", v );
```
```suggestion
    printf( "range: %lf\n", v );
```

kgryte reviewed (changes requested):
Given the copy-paste errors in the README, may be worth taking another look at this PR before further review.

aayush0325 commented:
> Given the copy-paste errors in the README, may be worth taking another look at this PR before further review.

yes, taking care of this here and in other open PRs

aayush0325 committed:
chore: stuff from code review

aayush0325 committed:
chore: missing changes

aayush0325 commented:
i think everything else should be okay for now here, the javascript benchmarks need to be looked at to ensure if i've taken the correct approach to avoid using manual for loops and using stdlib dependencies instead.

im getting this linting error in the tests
```bash
Linting file: lib/node_modules/@stdlib/stats/base/dmskrange/README.md
lib/node_modules/@stdlib/stats/base/dmskrange/README.md
  215:1-215:96  error  Use headings shorter than `80`  maximum-heading-length  remark-lint
``` 

how can i resolve this? @kgryte. i'll make similar changes to the other open PRs once this one gets approved

aayush0325 commented:
`<!-- remark-lint-disable maximum-heading-length -->`

i found this after looking around a bit, this disables the max heading lint for the whole file though. not sure if i should use this

kgryte commented:
@aayush0325 Yes, that is fine.

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/benchmark/benchmark.js:
```diff
@@ -21,17 +21,37 @@
+* @returns {number} one or zero
+*/
+function rand() {
+	if ( bernoulli( 0.2 ) ) {
```
```suggestion
	if ( bernoulli( 0.2 ) < 1 ) {
```

My  preference is to make the comparison explicit.

kgryte replied:
But also, you don't need to do this. Just use `random/array/bernoulli` directly. A Bernoulli RV is either `0` or `1`, which is exactly what you want for the mask arrays.

kgryte replied:
E.g., `bernoulli( len, 0.2, boptions )` where `boptions = { 'dtype': 'uint8' }`.

aayush0325 replied:
ah well, there's something new to learn! pushing the required changes soon!

aayush0325 committed:
feat: update benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: update readme to show mask is an array

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

aayush0325 commented:
i've updated the benchmarks, although still not able to fix the linting error, need some help there @kgryte.

aayush0325 committed:
chore: typo in type name

aayush0325 committed:
feat: update examples

aayush0325 commented:
i think by making use of `random/array/bernoulli` we can refactor the javascript examples too as done in previous packages, i've given it a try please go over that too @kgryte

aayush0325 commented:
just this error left to resolve now

```bash
Linting file: lib/node_modules/@stdlib/stats/base/dmskrange/README.md
lib/node_modules/@stdlib/stats/base/dmskrange/README.md
  215:1-215:96  error  Use headings shorter than `80`  maximum-heading-length  remark-lint
```

kgryte commented:
@aayush0325 Did `<!-- remark-lint-disable maximum-heading-length -->` not work?

aayush0325 commented:
> @aayush0325 Did `<!-- remark-lint-disable maximum-heading-length -->` not work?

tried it, wasnt able to commit

kgryte commented:
Also, it should be `<!-- lint disable maximum-heading-length -->`

aayush0325 commented:
> Also, it should be `<!-- lint disable maximum-heading-length -->`

ah lemme try this

aayush0325 committed:
chore: resolve lint error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 commented:
works @kgryte, pushed

aayush0325 commented:
do we need anything else @kgryte?

aayush0325 committed:
chore: comma spacing fix

aayush0325 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/README.md:
```diff
@@ -117,18 +111,16 @@ The function has the following additional parameters:
 -   **offsetMask**: starting index for `mask`.
 
-While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on a starting index. For example, to calculate the [range][range] for every other value in `x` starting from the second value
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on a starting index. For example, to calculate the [range][range] for every other element in `x` starting from the second element
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to calculate the [range][range] for every other element in `x` starting from the second element
```

In this case, because we have more than one offset, we need to use the plural forms consistently.

kgryte committed:
docs: fix grammar

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/dmskrange/docs/repl.txt:
```diff
@@ -3,8 +3,8 @@
 
-    The `N` and `stride` parameters determine which elements are accessed at
-    runtime.
+    The `N` and stride parameters determine which elements in the strided array
+    are accessed at runtime.
```
```suggestion
    The `N` and stride parameters determine which elements in the strided arrays
    are accessed at runtime.
```

kgryte committed:
docs: update note

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM. Thanks for working on this, @aayush0325!
